activerecord/lib/active_record/insert_all.rb

```diff
@@ -2,13 +2,18 @@
 
 module ActiveRecord
   class InsertAll # :nodoc:
-    attr_reader :model, :connection, :inserts, :keys
+    attr_reader :model, :connection, :inserts, :keys, :scope_for_create
     attr_reader :on_duplicate, :returning, :unique_by
 
-    def initialize(model, inserts, on_duplicate:, returning: nil, unique_by: nil)
+    def initialize(scope, inserts, on_duplicate:, returning: nil, unique_by: nil)
+      raise ArgumentError, "Bulk insert is currently not supported for through relations" if scope.try(:proxy_association).try(:reflection).try(:through_reflection?)
```

Comment:
It is not the best way how to check for `through` relation, but I wasn't able to find better for now.